### PR TITLE
Restore allocate_varlen_buffer

### PR DESCRIPTION
**Pull request by guilhermeleobas**
Solves gh-87

In this PR we're doing two things:
1. Remove target_info argument from RemoteJIT* classes
2. Hackish way to prevent numba from calling _ensure_finalize
3. Replace "calloc" by "allocate_varlen_buffer"

~PR is currently segfaulting in test `test_omnisci_math::positive`:~

```
~pytest rbc/tests/test_omnisci_math.py -k 'positive'~
```

cc @pearu

### CODE REVIEW

**Comment by guilhermeleobas** [comment]
Circle CI failure is not related to any changes introduced in this PR

File: rbc/irtools.py [hunk 1]
```diff
@@ -140,18 +167,17 @@ def init(self):
...
 
-    def __init__(self, typing_context, target_info):
-        self.target_info = target_info
+    def __init__(self, typing_context):
         super().__init__(typing_context)
 
     @compiler_lock.global_compiler_lock
     def init(self):
-        self.address_size = self.target_info.bits
+        target_info = TargetInfo()
```
Comment (guilhermeleobas) [inline rbc/irtools.py:177]: I guess this is not required anymore

File: rbc/irtools.py [hunk 2]
```diff
@@ -104,25 +125,31 @@ def _get_host_cpu_features(self):
...
-    def __init__(self, typing_context, target_info):
-        self.target_info = target_info
-        super(RemoteCPUContext, self).__init__(typing_context)
+    def __init__(self, typing_context):
+        super(JITRemoteCPUContext, self).__init__(typing_context)
 
     @compiler_lock.global_compiler_lock
     def init(self):
-        self.address_size = self.target_info.bits
+        target_info = TargetInfo()
```
Comment (guilhermeleobas) [inline rbc/irtools.py:146]: same

File: rbc/irtools.py [hunk 3]
```diff
@@ -356,84 +375,85 @@ def compile_to_LLVM(functions_and_signatures,
...
+        if target_info.is_cpu:
             typing_context = typing.Context()
-            target_context = RemoteCPUContext(typing_context, target)
-        elif target.is_gpu:
-            typing_context = RemoteGPUTypingContext()
-            target_context = RemoteGPUTargetContext(typing_context, target)
+            target_context = JITRemoteCPUContext(typing_context)
+        elif target_info.is_gpu:
+            typing_context = JITRemoteGPUTypingContext()
+            target_context = JITRemoteGPUTargetContext(typing_context)
```
Comment (pearu) [inline rbc/irtools.py:390]: It looks like we can eliminate `target_info.is_gpu` block here and merge `JITRemoteCPUContext` and `JITRemoteGPUTargetContext` to `JITRemoteTargetContext` to simplify irtools.py code a little.
After all, for GPU LLVM IR we are using numba CPU code generation with minor adjustments.

File: rbc/irtools.py (hunk 1 above)
Comment (pearu) [reply rbc/irtools.py:160]: I am not sure what do you mean but I'd keep `is32bit` and `address_size` to correspond to the bitness of the target: some GPU cards are 32-bit while the server runs 64-bit linux.

**Review by pearu — APPROVED** [review]
LGTM. I rebased the PR with master and did some code clean up.

I am fine with merging this PR as it is but I believe that irtools.py could be simplified a little, see my comments.

File: rbc/irtools.py (hunk 3 above)
Comment (guilhermeleobas) [reply rbc/irtools.py:390]: Actually, we need the `if-else` block to map math functions to libdevice when the target has GPU enabled. See:
https://github.com/xnd-project/rbc/blob/a34604c848250b670a79b214effffb8546ccdc7a/rbc/irtools.py#L182-L189

**Comment by guilhermeleobas** [comment]
>I am not sure what do you mean but I'd keep is32bit and address_size to correspond to the bitness of the target: some GPU cards are 32-bit while the server runs 64-bit linux.

It is still present, I just removed `target_info` from the list of arguments. See:
https://github.com/xnd-project/rbc/blob/a34604c848250b670a79b214effffb8546ccdc7a/rbc/irtools.py#L160-L163

**Pull request merged**